File: Stripe/STPSource.m
//
//  STPSource.m
//  Stripe
//
//  Created by Ben Guo on 1/23/17.
//  Copyright © 2017 Stripe, Inc. All rights reserved.
//


#import "STPSource.h"
#import "STPSource+Private.h"

#import "STPImageLibrary.h"
#import "STPLocalizationUtils.h"
#import "STPSourceOwner.h"
#import "STPSourceReceiver.h"
#import "STPSourceRedirect.h"
#import "STPSourceVerification.h"
#import "STPSourceWeChatPayDetails.h"

#import "NSDictionary+Stripe.h"

@interface STPSource ()

@property (nonatomic, nonnull) NSString *stripeID;
@property (nonatomic, nullable) NSNumber *amount;
@property (nonatomic, nullable) NSString *clientSecret;
@property (nonatomic, nullable) NSDate *created;
@property (nonatomic, nullable) NSString *currency;
@property (nonatomic) STPSourceFlow flow;
@property (nonatomic) BOOL livemode;
@property (nonatomic, copy, nullable, readwrite) NSDictionary<NSString *, NSString *> *metadata;
@property (nonatomic, nullable) STPSourceOwner *owner;
@property (nonatomic, nullable) STPSourceReceiver *receiver;
@property (nonatomic, nullable) STPSourceRedirect *redirect;
@property (nonatomic) STPSourceStatus status;
@property (nonatomic) STPSourceType type;
@property (nonatomic) STPSourceUsage usage;
@property (nonatomic, nullable) STPSourceVerification *verification;
@property (nonatomic, nullable) NSDictionary *details;
@property (nonatomic, nullable) STPSourceCardDetails *cardDetails;
@property (nonatomic, nullable) STPSourceSEPADebitDetails *sepaDebitDetails;
@property (nonatomic, nullable) STPSourceWeChatPayDetails *weChatPayDetails;
@property (nonatomic, readwrite, nonnull, copy) NSDictionary *allResponseFields;

// See STPSource+Private.h

@end

@implementation STPSource

#pragma mark - STPSourceType

+ (NSDictionary<NSString *,NSNumber *> *)stringToTypeMapping {
    return @{
             @"bancontact": @(STPSourceTypeBancontact),
             @"card": @(STPSourceTypeCard),
             @"giropay": @(STPSourceTypeGiropay),
             @"ideal": @(STPSourceTypeIDEAL),
             @"sepa_debit": @(STPSourceTypeSEPADebit),
             @"sofort": @(STPSourceTypeSofort),
             @"three_d_secure": @(STPSourceTypeThreeDSecure),
             @"alipay": @(STPSourceTypeAlipay),
             @"p24": @(STPSourceTypeP24),
             @"eps": @(STPSourceTypeEPS),
             @"multibanco": @(STPSourceTypeMultibanco),
             @"wechat": @(STPSourceTypeWeChatPay),
             };
}

+ (STPSourceType)typeFromString:(NSString *)string {
    NSString *key = [string lowercaseString];
    NSNumber *typeNumber = [self stringToTypeMapping][key];

    if (typeNumber != nil) {
        return (STPSourceType)[typeNumber integerValue];
    }

    return STPSourceTypeUnknown;
}

+ (nullable NSString *)stringFromType:(STPSourceType)type {
    return [[[self stringToTypeMapping] allKeysForObject:@(type)] firstObject];
}

#pragma mark - STPSourceFlow

+ (NSDictionary<NSString *,NSNumber *> *)stringToFlowMapping {
    return @{
             @"redirect": @(STPSourceFlowRedirect),
             @"receiver": @(STPSourceFlowReceiver),
             @"code_verification": @(STPSourceFlowCodeVerification),
             @"none": @(STPSourceFlowNone),
             };
}

+ (STPSourceFlow)flowFromString:(NSString *)string {
    NSString *key = [string lowercaseString];
    NSNumber *flowNumber = [self stringToFlowMapping][key];

    if (flowNumber != nil) {
        return (STPSourceFlow)[flowNumber integerValue];
    }

    return STPSourceFlowUnknown;
}

+ (nullable NSString *)stringFromFlow:(STPSourceFlow)flow {
    return [[[self stringToFlowMapping] allKeysForObject:@(flow)] firstObject];
}

#pragma mark - STPSourceStatus

+ (NSDictionary <NSString *, NSNumber *> *)stringToStatusMapping {
    return @{
             @"pending": @(STPSourceStatusPending),
             @"chargeable": @(STPSourceStatusChargeable),
             @"consumed": @(STPSourceStatusConsumed),
             @"canceled": @(STPSourceStatusCanceled),
             @"failed": @(STPSourceStatusFailed),
             };
}

+ (STPSourceStatus)statusFromString:(NSString *)string {
    NSString *key = [string lowercaseString];
    NSNumber *statusNumber = [self stringToStatusMapping][key];

    if (statusNumber != nil) {
        return (STPSourceStatus)[statusNumber integerValue];
    }

    return STPSourceStatusUnknown;
}

+ (nullable NSString *)stringFromStatus:(STPSourceStatus)status {
    return [[[self stringToStatusMapping] allKeysForObject:@(status)] firstObject];
}

#pragma mark - STPSourceUsage

+ (NSDictionary<NSString *,NSNumber *> *)stringToUsageMapping {
    return @{
             @"reusable": @(STPSourceUsageReusable),
             @"single_use": @(STPSourceUsageSingleUse),
             };
}

+ (STPSourceUsage)usageFromString:(NSString *)string {
    NSString *key = [string lowercaseString];
    NSNumber *usageNumber = [self stringToUsageMapping][key];

    if (usageNumber != nil) {
        return (STPSourceUsage)[usageNumber integerValue];
    }

    return STPSourceUsageUnknown;
}

+ (nullable NSString *)stringFromUsage:(STPSourceUsage)usage {
    return [[[self stringToUsageMapping] allKeysForObject:@(usage)] firstObject];
}

#pragma mark - Equality

- (BOOL)isEqual:(STPSource *)source {
    return [self isEqualToSource:source];
}

- (NSUInteger)hash {
    return [self.stripeID hash];
}

- (BOOL)isEqualToSource:(STPSource *)source {
    if (self == source) {
        return YES;
    }

    if (!source || ![source isKindOfClass:self.class]) {
        return NO;
    }

    return [self.stripeID isEqualToString:source.stripeID];
}

#pragma mark - Description

- (NSString *)description {
    NSArray *props = @[
                       // Object
                       [NSString stringWithFormat:@"%@: %p", NSStringFromClass([self class]), self],

                       // Identifier
                       [NSString stringWithFormat:@"stripeID = %@", self.stripeID],

                       // Source details (alphabetical)
                       [NSString stringWithFormat:@"amount = %@", self.amount],
                       [NSString stringWithFormat:@"clientSecret = %@", (self.clientSecret) ? @"<redacted>" : nil],
                       [NSString stringWithFormat:@"created = %@", self.created],
                       [NSString stringWithFormat:@"currency = %@", self.currency],
                       [NSString stringWithFormat:@"flow = %@", ([self.class stringFromFlow:self.flow]) ?: @"unknown"],
                       [NSString stringWithFormat:@"livemode = %@", (self.livemode) ? @"YES" : @"NO"],
                       [NSString stringWithFormat:@"metadata = %@", (self.metadata) ? @"<redacted>" : nil],
                       [NSString stringWithFormat:@"owner = %@", (self.owner) ? @"<redacted>" : nil],
                       [NSString stringWithFormat:@"receiver = %@", self.receiver],
                       [NSString stringWithFormat:@"redirect = %@", self.redirect],
                       [NSString stringWithFormat:@"status = %@", ([self.class stringFromStatus:self.status]) ?: @"unknown"],
                       [NSString stringWithFormat:@"type = %@", ([self.class stringFromType:self.type]) ?: @"unknown"],
                       [NSString stringWithFormat:@"usage = %@", ([self.class stringFromUsage:self.usage]) ?: @"unknown"],
                       [NSString stringWithFormat:@"verification = %@", self.verification],
                       ];

    return [NSString stringWithFormat:@"<%@>", [props componentsJoinedByString:@"; "]];
}

#pragma mark - STPAPIResponseDecodable

- (NSString *)stripeObject {
    return @"source";
}

+ (instancetype)decodedObjectFromAPIResponse:(NSDictionary *)response {
    NSDictionary *dict = [response stp_dictionaryByRemovingNulls];
    if (!dict) {
        return nil;
    }

    // required fields
    NSString *stripeId = [dict stp_stringForKey:@"id"];
    NSString *rawStatus = [dict stp_stringForKey:@"status"];
    NSString *rawType = [dict stp_stringForKey:@"type"];
    if (!stripeId || !rawStatus || !rawType || !dict[@"livemode"]) {
        return nil;
    }

    STPSource *source = [self new];
    source.stripeID = stripeId;
    source.amount = [dict stp_numberForKey:@"amount"];
    source.clientSecret = [dict stp_stringForKey:@"client_secret"];
    source.created = [dict stp_dateForKey:@"created"];
    source.currency = [dict stp_stringForKey:@"currency"];
    NSString *rawFlow = [dict stp_stringForKey:@"flow"];
    source.flow = [[self class] flowFromString:rawFlow];
    source.livemode = [dict stp_boolForKey:@"livemode" or:YES];
    source.metadata = [[dict stp_dictionaryForKey:@"metadata"] stp_dictionaryByRemovingNonStrings];
    NSDictionary *rawOwner = [dict stp_dictionaryForKey:@"owner"];
    source.owner = [STPSourceOwner decodedObjectFromAPIResponse:rawOwner];
    NSDictionary *rawReceiver = [dict stp_dictionaryForKey:@"receiver"];
    source.receiver = [STPSourceReceiver decodedObjectFromAPIResponse:rawReceiver];
    NSDictionary *rawRedirect = [dict stp_dictionaryForKey:@"redirect"];
    source.redirect = [STPSourceRedirect decodedObjectFromAPIResponse:rawRedirect];
    source.status = [[self class] statusFromString:rawStatus];
    source.type = [[self class] typeFromString:rawType];
    NSString *rawUsage = [dict stp_stringForKey:@"usage"];
    source.usage = [[self class] usageFromString:rawUsage];
    NSDictionary *rawVerification = [dict stp_dictionaryForKey:@"verification"];
    if (rawVerification) {
        source.verification = [STPSourceVerification decodedObjectFromAPIResponse:rawVerification];
    }
    source.details = [dict stp_dictionaryForKey:rawType];
    source.allResponseFields = dict;

    if (source.type == STPSourceTypeCard) {
        source.cardDetails = [STPSourceCardDetails decodedObjectFromAPIResponse:source.details];
    } else if (source.type == STPSourceTypeSEPADebit) {
        source.sepaDebitDetails = [STPSourceSEPADebitDetails decodedObjectFromAPIResponse:source.details];
    } else if (source.type == STPSourceTypeWeChatPay) {
        source.weChatPayDetails = [STPSourceWeChatPayDetails decodedObjectFromAPIResponse:source.details];
    }

    return source;
}

#pragma mark - STPPaymentOption

- (UIImage *)image {
    if (self.type == STPSourceTypeCard && self.cardDetails != nil) {
        return [STPImageLibrary brandImageForCardBrand:self.cardDetails.brand];
    } else {
        return [STPImageLibrary brandImageForCardBrand:STPCardBrandUnknown];
    }
}

- (UIImage *)templateImage {
    if (self.type == STPSourceTypeCard && self.cardDetails != nil) {
        return [STPImageLibrary templatedBrandImageForCardBrand:self.cardDetails.brand];
    } else {
        return [STPImageLibrary templatedBrandImageForCardBrand:STPCardBrandUnknown];
    }
}

- (NSString *)label {
    switch (self.type) {
        case STPSourceTypeBancontact:
            return STPLocalizedString(@"Bancontact", @"Source type brand name");
        case STPSourceTypeCard:
            if (self.cardDetails != nil) {
                NSString *brand = [STPCard stringFromBrand:self.cardDetails.brand];
                return [NSString stringWithFormat:@"%@ %@", brand, self.cardDetails.last4];
            } else {
                return [STPCard stringFromBrand:STPCardBrandUnknown];
            }
        case STPSourceTypeGiropay:
            return STPLocalizedString(@"Giropay", @"Source type brand name");
        case STPSourceTypeIDEAL:
            return STPLocalizedString(@"iDEAL", @"Source type brand name");
        case STPSourceTypeSEPADebit:
            return STPLocalizedString(@"SEPA Direct Debit", @"Source type brand name");
        case STPSourceTypeSofort:
            return STPLocalizedString(@"SOFORT", @"Source type brand name");
        case STPSourceTypeThreeDSecure:
            return STPLocalizedString(@"3D Secure", @"Source type brand name");
        case STPSourceTypeAlipay:
            return STPLocalizedString(@"Alipay", @"Source type brand name");
        case STPSourceTypeP24:
            return STPLocalizedString(@"P24", @"Source type brand name");
        case STPSourceTypeEPS:
            return STPLocalizedString(@"EPS", @"Source type brand name");
        case STPSourceTypeMultibanco:
            return STPLocalizedString(@"Multibanco", @"Source type brand name");
        case STPSourceTypeWeChatPay:
            return STPLocalizedString(@"WeChat Pay", @"Source type brand name");
        case STPSourceTypeUnknown:
            return STPLocalizedString(@"Unknown", @"Default missing source type label");
    }
}

- (BOOL)isReusable {
    return (self.usage != STPSourceUsageSingleUse);
}

@end

File: StripePayments/StripePayments/Source/API Bindings/Models/Sources/STPSource.swift
//
//  STPSource.swift
//  StripePayments
//
//  Created by Ben Guo on 1/23/17.
//  Copyright © 2017 Stripe, Inc. All rights reserved.
//

import Foundation
import UIKit

/// Representation of a customer's payment instrument created with the Stripe API. - seealso: https://stripe.com/docs/api#sources
public class STPSource: NSObject, STPAPIResponseDecodable, STPSourceProtocol {

    /// You cannot directly instantiate an `STPSource`. You should only use one that
    /// has been returned from an `STPAPIClient` callback.
    override required init() {
        super.init()
    }

    /// The amount associated with the source.
    @objc public private(set) var amount: NSNumber?
    /// The client secret of the source. Used for client-side fetching of a source
    /// using a publishable key.
    @objc public private(set) var clientSecret: String?
    /// When the source was created.
    @objc public private(set) var created: Date = Date(timeIntervalSince1970: TimeInterval(0))
    /// The currency associated with the source.
    @objc public private(set) var currency: String?
    /// Whether or not this source was created in livemode.
    @objc public private(set) var livemode = false
    /// The status of the source.
    @objc public private(set) var status: STPSourceStatus = .unknown
    /// The type of the source.
    @objc public private(set) var type: STPSourceType = .unknown
    /// Whether this source should be reusable or not.
    @objc public private(set) var usage: STPSourceUsage = .unknown
    /// Information about the source specific to its type
    @objc public private(set) var details: [AnyHashable: Any]?
    /// If this is a card source, this property provides typed access to the
    /// contents of the `details` dictionary.
    @objc public private(set) var cardDetails: STPSourceCardDetails?
    // MARK: - Deprecated

    /// A set of key/value pairs associated with the source object.
    /// @deprecated Metadata is no longer returned to clients using publishable keys. Retrieve them on your server using yoursecret key instead.
    /// - seealso: https://stripe.com/docs/api#metadata
    @available(
        *,
        deprecated,
        message:
            "Metadata is no longer returned to clients using publishable keys. Retrieve them on your server using yoursecret key instead."
    )
    @objc public private(set) var metadata: [String: String]?
    @objc public var stripeID = ""
    @objc private(set) public var allResponseFields: [AnyHashable: Any] = [:]

    // MARK: - STPSourceType
    class func stringToTypeMapping() -> [String: NSNumber] {
        return [
            "card": NSNumber(value: STPSourceType.card.rawValue),
        ]
    }

    @objc(typeFromString:)
    class func type(from string: String) -> STPSourceType {
        let key = string.lowercased()
        let typeNumber = self.stringToTypeMapping()[key]

        if let typeNumber = typeNumber {
            return (STPSourceType(rawValue: typeNumber.intValue))!
        }

        return .unknown
    }

    @objc(stringFromType:)
    class func string(from type: STPSourceType) -> String? {
        return
            (self.stringToTypeMapping() as NSDictionary).allKeys(
                for: NSNumber(value: type.rawValue)
            )
            .first as? String
    }

    // MARK: - STPSourceStatus
    class func stringToStatusMapping() -> [String: NSNumber] {
        return [
            "pending": NSNumber(value: STPSourceStatus.pending.rawValue),
            "chargeable": NSNumber(value: STPSourceStatus.chargeable.rawValue),
            "consumed": NSNumber(value: STPSourceStatus.consumed.rawValue),
            "canceled": NSNumber(value: STPSourceStatus.canceled.rawValue),
            "failed": NSNumber(value: STPSourceStatus.failed.rawValue),
        ]
    }

    @objc(statusFromString:)
    class func status(from string: String) -> STPSourceStatus {
        let key = string.lowercased()
        let statusNumber = self.stringToStatusMapping()[key]

        if let statusNumber = statusNumber {
            return (STPSourceStatus(rawValue: statusNumber.intValue))!
        }

        return .unknown
    }

    @objc(stringFromStatus:)
    class func string(from status: STPSourceStatus) -> String? {
        return
            (self.stringToStatusMapping() as NSDictionary).allKeys(
                for: NSNumber(value: status.rawValue)
            )
            .first as? String
    }

    // MARK: - STPSourceUsage
    class func stringToUsageMapping() -> [String: NSNumber] {
        return [
            "reusable": NSNumber(value: STPSourceUsage.reusable.rawValue),
            "single_use": NSNumber(value: STPSourceUsage.singleUse.rawValue),
        ]
    }

    @objc(usageFromString:)
    class func usage(from string: String) -> STPSourceUsage {
        let key = string.lowercased()
        let usageNumber = self.stringToUsageMapping()[key]

        if let usageNumber = usageNumber {
            return (STPSourceUsage(rawValue: usageNumber.intValue))!
        }

        return .unknown
    }

    @objc(stringFromUsage:)
    class func string(from usage: STPSourceUsage) -> String? {
        return
            (self.stringToUsageMapping() as NSDictionary).allKeys(
                for: NSNumber(value: usage.rawValue)
            )
            .first as? String
    }

    // MARK: - Equality
    /// :nodoc:
    @objc
    public override func isEqual(_ object: Any?) -> Bool {
        return isEqual(to: object as? STPSource)
    }

    /// :nodoc:
    @objc public override var hash: Int {
        return stripeID.hash
    }

    func isEqual(to source: STPSource?) -> Bool {
        if self === source {
            return true
        }

        guard let source = source else {
            return false
        }

        return stripeID == source.stripeID
    }

    // MARK: - Description
    /// :nodoc:
    @objc public override var description: String {
        let props = [
            // Object
            String(format: "%@: %p", NSStringFromClass(STPSource.self), self),
            // Identifier
            "stripeID = \(stripeID)",
            // Source details (alphabetical)
            "amount = \(amount ?? 0)",
            "clientSecret = \(((clientSecret) != nil ? "<redacted>" : nil) ?? "")",
            "created = \(String(describing: created))",
            "currency = \(currency ?? "")",
            "livemode = \((livemode) ? "YES" : "NO")",
            "status = \((STPSource.string(from: status)) ?? "unknown")",
            "type = \((STPSource.string(from: type)) ?? "unknown")",
            "usage = \((STPSource.string(from: usage)) ?? "unknown")",
        ]

        return "<\(props.joined(separator: "; "))>"
    }

    // MARK: - STPAPIResponseDecodable
    @objc func stripeObject() -> String {
        return "source"
    }

    @objc
    public class func decodedObject(fromAPIResponse response: [AnyHashable: Any]?) -> Self? {
        guard let response = response else {
            return nil
        }
        let dict = response.stp_dictionaryByRemovingNulls()

        // required fields
        let stripeId = dict.stp_string(forKey: "id")
        let rawStatus = dict.stp_string(forKey: "status")
        let rawType = dict.stp_string(forKey: "type")
        if stripeId == nil || rawStatus == nil || rawType == nil || dict["livemode"] == nil {
            return nil
        }

        let source = self.init()
        source.stripeID = stripeId ?? ""
        source.amount = dict.stp_number(forKey: "amount")
        source.clientSecret = dict.stp_string(forKey: "client_secret")
        source.created = dict.stp_date(forKey: "created") ?? Date(timeIntervalSince1970: TimeInterval(0))
        source.currency = dict.stp_string(forKey: "currency")
        source.livemode = dict.stp_bool(forKey: "livemode", or: true)
        source.status = self.status(from: rawStatus ?? "")
        source.type = self.type(from: rawType ?? "")
        let rawUsage = dict.stp_string(forKey: "usage")
        source.usage = self.usage(from: rawUsage ?? "")
        source.details = dict.stp_dictionary(forKey: rawType ?? "")
        source.allResponseFields = dict

        if source.type == .card {
            if let details1 = source.details {
                source.cardDetails = STPSourceCardDetails.decodedObject(fromAPIResponse: details1)
            }
        }

        return source
    }
}
